app/scripts/lib/indexed-db-backend.ts

```diff
@@ -0,0 +1,185 @@
+  private storeName: string;
+
+  private dbVersion: number;
+
+  constructor(storeName: string, dbVersion: number) {
+    this.storeName = storeName;
+    this.dbVersion = dbVersion;
+  }
+
+  private async _getObjectStore(
+    mode: IDBTransactionMode,
+  ): Promise<IDBObjectStore> {
+    return new Promise((resolve, reject) => {
+      const request = indexedDB.open(this.storeName, this.dbVersion);
+
+      request.onerror = (event: Event) => {
+        reject(
+          new Error(
+            `Failed to open database ${this.storeName}: ${
+              (event.target as any)?.error
+            }`,
+          ),
+        );
+      };
+
+      request.onupgradeneeded = (event) => {
+        const db = (event.target as IDBOpenDBRequest).result;
+
+        if (!db.objectStoreNames.contains(this.storeName)) {
+          db.createObjectStore(this.storeName, {
+            keyPath: ['name', 'chainId'],
+          });
+        }
+      };
+
+      request.onsuccess = (event) => {
+        const db = (event.target as IDBOpenDBRequest).result;
+        const transaction = db.transaction([this.storeName], mode);
+        const objectStore = transaction.objectStore(this.storeName);
+        resolve(objectStore);
+      };
+    });
+  }
+
+  async read(key: StorageKey, checksum: string): Promise<ArrayBuffer> {
+    return new Promise<ArrayBuffer>((resolve, reject) => {
+      this._getObjectStore('readonly')
+        .then((objectStore) => {
+          const request = objectStore.get([key.name, key.chainId]);
+
+          request.onsuccess = async (event) => {
+            const data = (event.target as any)?.result?.data;
+            await validateChecksum(key, data, checksum).catch((error) => {
+              reject(error);
+            });
+            resolve(data);
+          };
+
+          request.onerror = (event) => {
+            reject(
+              new Error(`Error reading data: ${(event.target as any)?.error}`),
+            );
+          };
+        })
+        .catch((error) => {
+          reject(error);
+        });
+    });
+  }
+
+  async write(
+    key: StorageKey,
+    data: ArrayBuffer,
+    checksum: string,
+  ): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      validateChecksum(key, data, checksum)
+        .then(() => {
+          this._getObjectStore('readwrite')
+            .then((objectStore) => {
+              const request = objectStore.put({ ...key, data });
+
+              request.onsuccess = () => {
+                resolve();
+              };
+
+              request.onerror = (event) => {
+                reject(
+                  new Error(
+                    `Error writing data: ${(event.target as any)?.error}`,
+                  ),
+                );
+              };
+            })
+            .catch((error) => {
+              reject(error);
+            });
+        })
+        .catch((error) => {
+          reject(error);
+        });
```

Comment:
Instead of using Promise chainning we should use async/await wrapped in a Promise. IMO it improves the code readibility. Something like this:
```js
return new Promise<void>(async (resolve, reject) => {
  try {
    await validateChecksum(key, data, checksum);
    const objectuStore = await  this._getObjectStore('readwrite');
    const request = objectStore.put({ ...key, data });

    request.onsuccess = () => {
      resolve();
    };

    request.onerror = (event) => {
      reject(
        new Error(
          `Error writing data: ${(event.target as any)?.error}`,
        ),
      );
    };
  } catch (error) {
    reject(error);
  }
});
```
(the same pattern could be used on the other methods)

What do you think?

Reply:
@cryptotavares : this seems not valid eslint code. Please check https://eslint.org/docs/latest/rules/no-async-promise-executor